src/Aws/Lambda/Runtime/Common.hs

```diff
@@ -1,90 +1,73 @@
-{-# LANGUAGE DeriveLift                 #-}
-{-# LANGUAGE DerivingStrategies         #-}
-{-# LANGUAGE FlexibleInstances          #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
 {-# LANGUAGE GeneralizedNewtypeDeriving #-}
-{-# LANGUAGE UndecidableInstances       #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE UndecidableInstances #-}
 
 module Aws.Lambda.Runtime.Common
-  ( RunCallback
-  , LambdaResult(..)
-  , LambdaError(..)
-  , LambdaOptions(..)
-  , DispatcherOptions(..)
-  , ApiGatewayDispatcherOptions(..)
-  , DispatcherStrategy(..)
-  , ToLambdaResponseBody(..)
-  , unLambdaResponseBody
-  , defaultDispatcherOptions
-  ) where
+  ( RunCallback,
+    LambdaResult (..),
+    LambdaError (..),
+    LambdaOptions (..),
+    ApiGatewayDispatcherOptions (..),
+    HandlerType (..),
+    HandlerName (..),
+    RawEventObject,
+  )
+where
 
-import Aws.Lambda.Runtime.ApiGatewayInfo
+import Aws.Lambda.Runtime.ALB.Types
+import Aws.Lambda.Runtime.APIGateway.Types
+  ( ApiGatewayDispatcherOptions (..),
+    ApiGatewayResponse,
+    ApiGatewayResponseBody,
+  )
 import Aws.Lambda.Runtime.Context (Context)
-import Aws.Lambda.Utilities
-import Data.Aeson (FromJSON, ToJSON)
+import Aws.Lambda.Runtime.StandaloneLambda.Types
+  ( StandaloneLambdaResponseBody,
+  )
 import qualified Data.ByteString.Lazy as Lazy
+import Data.Hashable (Hashable)
 import Data.Text (Text)
-import qualified Data.Text as Text
 import GHC.Generics (Generic)
-import Language.Haskell.TH.Syntax (Lift)
-
--- | API Gateway specific dispatcher options
-newtype ApiGatewayDispatcherOptions = ApiGatewayDispatcherOptions
-  { propagateImpureExceptions :: Bool
-  -- ^ Should impure exceptions be propagated through the API Gateway interface
-  } deriving (Lift)
-
--- | Options that the dispatcher generator expects
-newtype DispatcherOptions = DispatcherOptions
-  { apiGatewayDispatcherOptions :: ApiGatewayDispatcherOptions
-  } deriving (Lift)
-
-defaultDispatcherOptions :: DispatcherOptions
-defaultDispatcherOptions =
-  DispatcherOptions (ApiGatewayDispatcherOptions True)
-
--- | A strategy on how to generate the dispatcher functions
-data DispatcherStrategy =
-  UseWithAPIGateway |
-  StandaloneLambda
-  deriving (Lift)
+import Data.String (IsString)
 
 -- | Callback that we pass to the dispatcher function
-type RunCallback context =
-  LambdaOptions context -> IO (Either LambdaError LambdaResult)
+type RunCallback (t :: HandlerType) context =
+  LambdaOptions context -> IO (Either (LambdaError t) (LambdaResult t))
```

Comment:
Oh, I see now where the `t` comes from 😅 

Still, even though it seems redundant, I'd call this like

```suggestion
type RunCallback (handlerType :: HandlerType) context =
  LambdaOptions context -> IO (Either (LambdaError handlerType) (LambdaResult handlerType))
```